File: frameworks/core/interfaces/arkoala/implementation/button_modifier.cpp
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "bridge/common/utils/utils.h"
#include "core/components/common/layout/constants.h"
#include "core/components/common/properties/color.h"
#include "core/components_ng/pattern/button/button_model_ng.h"
#include "core/components_ng/pattern/button/button_request_data.h"
#include "core/interfaces/native/node/node_api.h"
#include "arkoala_api_generated.h"
#include "core/interfaces/arkoala/utility/converter.h"

namespace OHOS::Ace::NG {
struct ButtonOptions {
    std::optional<ButtonType> type;
    std::optional<ButtonRole> role;
    std::optional<bool> stateEffect;
    std::optional<ControlSize> controlSize;
    std::optional<ButtonStyleMode> buttonStyle;
};
} // OHOS::Ace::NG

namespace OHOS::Ace::NG::Converter {
template<>
ButtonOptions Convert(const Ark_ButtonOptions& src)
{
    ButtonOptions options;
    options.type = OptConvert<ButtonType>(src.type);
    options.role = OptConvert<ButtonRole>(src.role);
    options.stateEffect = OptConvert<bool>(src.stateEffect);
    options.controlSize = OptConvert<ControlSize>(src.controlSize);
    options.buttonStyle = OptConvert<ButtonStyleMode>(src.buttonStyle);
    return options;
}
}

namespace OHOS::Ace::NG::GeneratedModifier {
namespace ButtonInterfaceModifier {
void SetButtonOptions0Impl(Ark_NativePointer node)
{
    // safe it empty for save default values
}
void SetButtonOptions1Impl(Ark_NativePointer node,
                           const Ark_ButtonOptions* options)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    CHECK_NULL_VOID(options);
    auto buttonOptions = Converter::Convert<ButtonOptions>(*options);
    if (buttonOptions.type) {
        ButtonModelNG::SetType(frameNode, static_cast<int>(buttonOptions.type.value()));
    }
    if (buttonOptions.stateEffect) {
        ButtonModelNG::SetStateEffect(frameNode, buttonOptions.stateEffect.value());
    }
    ButtonModelNG::SetRole(frameNode, buttonOptions.role);
    ButtonModelNG::SetControlSize(frameNode, buttonOptions.controlSize);
    ButtonModelNG::SetButtonStyle(frameNode, buttonOptions.buttonStyle);
}
void SetButtonOptions2Impl(Ark_NativePointer node,
                           const ResourceStr* label,
                           const Opt_ButtonOptions* options)
{
    if (options->tag == ARK_TAG_UNDEFINED) {
        SetButtonOptions0Impl(node);
    } else {
        SetButtonOptions1Impl(node, &options->value);
    }
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetLabel(frameNode, Converter::ConvertOrDefault(*label, ""));
}
} // ButtonInterfaceModifier
namespace ButtonAttributeModifier {
constexpr int32_t DEFAULT_FONT_SIZE = 12;
constexpr int32_t DEFAULT_MAX_LINES = 100;

void TypeImpl(Ark_NativePointer node,
              Ark_ButtonType value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetType(frameNode, (ArkUI_Int32)Converter::ConvertOrDefault(value, ButtonType::CAPSULE));
}
void StateEffectImpl(Ark_NativePointer node,
                     Ark_Boolean value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetStateEffect(frameNode, Converter::ConvertOrDefault(value, true));
}
void ButtonStyleImpl(Ark_NativePointer node,
                     Ark_ButtonStyleMode value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetButtonStyle(frameNode, Converter::ConvertOrDefault(value, ButtonStyleMode::EMPHASIZE));
}
void ControlSizeImpl(Ark_NativePointer node,
                     Ark_ControlSize value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetControlSize(frameNode, Converter::ConvertOrDefault(value, ControlSize::NORMAL));
}
void RoleImpl(Ark_NativePointer node,
              Ark_ButtonRole value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetRole(frameNode, Converter::ConvertOrDefault(value, ButtonRole::NORMAL));
}
void FontColorImpl(Ark_NativePointer node,
                   const ResourceColor* value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetFontColor(frameNode, Converter::ConvertOrDefault(*value, Color()));
}
void FontSizeImpl(Ark_NativePointer node,
                  const Ark_Length* value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetFontSize(frameNode, CalcDimension(value->value, (DimensionUnit)value->unit));
}
void FontWeightImpl(Ark_NativePointer node,
                    const Type_ButtonAttribute_fontWeight_Arg0* value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetFontWeight(frameNode, Converter::ConvertOrDefault(*value, FontWeight::NORMAL));
}
void FontStyleImpl(Ark_NativePointer node,
                   Ark_FontStyle value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetFontStyle(frameNode, (Ace::FontStyle)value);
}
void FontFamilyImpl(Ark_NativePointer node,
                    const Type_ButtonAttribute_fontFamily_Arg0* value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonModelNG::SetFontFamily(frameNode,
                                 Framework::ConvertStrToFontFamilies(Converter::ConvertOrDefault(*value, "")));
}
void ContentModifierImpl(Ark_NativePointer node,
                         const Ark_CustomObject* modifier)
{
    LOGE("ARKOALA ButtonAttribute::ContentModifierImpl -> Method is not "
            "implemented.");
}
void LabelStyleImpl(Ark_NativePointer node,
                    const Ark_LabelStyle* value)
{
    auto frameNode = reinterpret_cast<FrameNode *>(node);
    CHECK_NULL_VOID(frameNode);
    ButtonParameters parameters;
    parameters.textOverflow = Converter::ConvertOrDefault(value->overflow, TextOverflow::ELLIPSIS);
    parameters.maxLines = Converter::ConvertOrDefault(value->maxLines, DEFAULT_MAX_LINES);
    parameters.heightAdaptivePolicy = Converter::ConvertOrDefault(
        value->heightAdaptivePolicy, TextHeightAdaptivePolicy::MAX_LINES_FIRST);
    parameters.minFontSize = Converter::ConvertOrDefault(value->minFontSize, Dimension());
    parameters.maxFontSize = Converter::ConvertOrDefault(value->maxFontSize, Dimension());

    parameters.fontSize = Dimension(DEFAULT_FONT_SIZE, DimensionUnit::VP);
    parameters.fontStyle = Ace::FontStyle::NORMAL;
    parameters.fontWeight = FontWeight::NORMAL;
    parameters.fontFamily = std::vector<std::string>();
    Converter::WithOptional(value->font, [&](const Ark_Font &font) {
        parameters.fontSize = Converter::ConvertOrDefault(font.size, parameters.fontSize.value());
        parameters.fontStyle = Converter::ConvertOrDefault(font.style, parameters.fontStyle.value());
        parameters.fontWeight = Converter::ConvertOrDefault(font.weight, parameters.fontWeight.value());
        parameters.fontFamily = Converter::ConvertOrDefault(font.family, parameters.fontFamily.value());
    });
    ButtonModelNG::SetLabelStyle(frameNode, parameters);
}
} // ButtonAttributeModifier

const GENERATED_ArkUIButtonModifier* GetButtonModifier()
{
    static const GENERATED_ArkUIButtonModifier ArkUIButtonModifierImpl {
        ButtonInterfaceModifier::SetButtonOptions0Impl,
        ButtonInterfaceModifier::SetButtonOptions1Impl,
        ButtonInterfaceModifier::SetButtonOptions2Impl,
        ButtonAttributeModifier::TypeImpl,
        ButtonAttributeModifier::StateEffectImpl,
        ButtonAttributeModifier::ButtonStyleImpl,
        ButtonAttributeModifier::ControlSizeImpl,
        ButtonAttributeModifier::RoleImpl,
        ButtonAttributeModifier::FontColorImpl,
        ButtonAttributeModifier::FontSizeImpl,
        ButtonAttributeModifier::FontWeightImpl,
        ButtonAttributeModifier::FontStyleImpl,
        ButtonAttributeModifier::FontFamilyImpl,
        ButtonAttributeModifier::ContentModifierImpl,
        ButtonAttributeModifier::LabelStyleImpl,
    };
    return &ArkUIButtonModifierImpl;
}
}

File: frameworks/bridge/declarative_frontend/ark_modifier/src/button_modifier.ts
/*
 * Copyright (c) 2024-2026 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class LazyArkButtonComponent extends ArkComponent {
  static module: ButtonComponentModule | undefined = undefined;

  constructor(nativePtr: KNode, classType: ModifierType) {
    super(nativePtr, classType);
    if (LazyArkButtonComponent.module === undefined) {
      LazyArkButtonComponent.module = globalThis.requireNapi('arkui.components.arkbutton');
    }
    this.lazyComponent = LazyArkButtonComponent.module.createComponent(nativePtr, classType);
  }

  setMap(): void {
    this.lazyComponent._modifiersWithKeys = this._modifiersWithKeys;
  }

  allowChildCount(): number {
    return this.lazyComponent.allowChildCount();
  }

  initialize(value: Object[]): this {
    this.lazyComponent.initialize(value);
    return this;
  }

  backgroundColor(value: ResourceColor): this {
    this.lazyComponent.backgroundColor(value);
    return this;
  }

  type(value: ButtonType): this {
    this.lazyComponent.type(value);
    return this;
  }

  stateEffect(value: boolean): this {
    this.lazyComponent.stateEffect(value);
    return this;
  }

  fontColor(value: ResourceColor): this {
    this.lazyComponent.fontColor(value);
    return this;
  }

  fontSize(value: Length): this {
    this.lazyComponent.fontSize(value);
    return this;
  }

  fontWeight(value: string | number | FontWeight): this {
    this.lazyComponent.fontWeight(value);
    return this;
  }

  fontStyle(value: FontStyle): this {
    this.lazyComponent.fontStyle(value);
    return this;
  }

  fontFamily(value: string | Resource): this {
    this.lazyComponent.fontFamily(value);
    return this;
  }

  labelStyle(value: LabelStyle): this {
    this.lazyComponent.labelStyle(value);
    return this;
  }

  borderRadius(value: Length | BorderRadiuses): this {
    this.lazyComponent.borderRadius(value);
    return this;
  }

  border(value: BorderOptions): this {
    this.lazyComponent.border(value);
    return this;
  }

  size(value: SizeOptions): this {
    this.lazyComponent.size(value);
    return this;
  }

  contentModifier(value: ContentModifier<ButtonConfiguration>): this {
    this.lazyComponent.contentModifier(value);
    return this;
  }

  role(value: ButtonRole): this {
    this.lazyComponent.role(value);
    return this;
  }

  buttonStyle(value: ButtonStyleMode): this {
    this.lazyComponent.buttonStyle(value);
    return this;
  }

  controlSize(value: ControlSize): this {
    this.lazyComponent.controlSize(value);
    return this;
  }

  minFontScale(value: number | Resource): this {
    this.lazyComponent.minFontScale(value);
    return this;
  }

  maxFontScale(value: number | Resource): this {
    this.lazyComponent.maxFontScale(value);
    return this;
  }

  fontFeature(value: string): this {
    return this;
  }

  minFontSize(value: number | string | Resource): this {
    return this;
  }

  maxFontSize(value: number | string | Resource): this {
    return this;
  }

  onClick(callback: (event: ClickEvent, result: ButtonCallbackResult) => void): this {
    this.lazyComponent.onClick(callback);
    return this;
  }
}

class ButtonModifier extends LazyArkButtonComponent implements AttributeModifier<ButtonAttribute> {

  constructor(nativePtr: KNode, classType: ModifierType) {
    super(nativePtr, classType);
    this._modifiersWithKeys = new ModifierMap();
    this.setMap();
  }

  applyNormalAttribute(instance: ButtonAttribute): void {
    ModifierUtils.applySetOnChange(this);
    ModifierUtils.applyAndMergeModifier<ButtonAttribute, ArkButtonComponent, ArkComponent>(instance, this);
  }
}
